python/chemiscope/structures/_shapes.py

```diff
@@ -0,0 +1,180 @@
+            [[0, i, i + 1] for i in range(1, n_points)]
+            + [[0, n_points, 1]]
+            + [[i, i + 1, i + 1 + n_points] for i in range(1, n_points)]
+            + [[n_points, 1, n_points + 1]]
+            + [[i + n_points, i, i + 1 + n_points] for i in range(1, n_points)]
+            + [[2 * n_points, n_points, n_points + 1]]
+            + [[i, i + 1, i + 1 + n_points] for i in range(1 + n_points, 2 * n_points)]
+            + [[2 * n_points, 1 + n_points, 2 * n_points + 1]]
+            + [
+                [i + n_points, i, i + 1 + n_points]
+                for i in range(1 + n_points, 2 * n_points)
+            ]
+            + [[3 * n_points, 2 * n_points, 2 * n_points + 1]]
+            + [
+                [3 * n_points + 1, i, i + 1]
+                for i in range(1 + 2 * n_points, 3 * n_points)
+            ]
+            + [[3 * n_points + 1, 3 * n_points, 2 * n_points + 1]]
+        ),
+    )
+    return arrow
+
+
+def _rotation_to_quaternion(R):
+    """Computes a quaternion that describes a rotation given the
+    rotation matrix R"""
+
+    m00, m01, m02 = R[0]
+    m10, m11, m12 = R[1]
+    m20, m21, m22 = R[2]
+    if m22 < 0:
+        if m00 > m11:
+            t = 1 + m00 - m11 - m22
+            q = np.array((t, m01 + m10, m20 + m02, m12 - m21))
+        else:
+            t = 1 - m00 + m11 - m22
+            q = np.array((m01 + m10, t, m12 + m21, m20 - m02))
+    else:
+        if m00 < -m11:
+            t = 1 - m00 - m11 + m22
+            q = np.array((m20 + m02, m12 + m21, t, m01 - m10))
+        else:
+            t = 1 + m00 + m11 + m22
+            q = np.array((m12 - m21, m20 - m02, m01 - m10, t))
+
+    q /= np.sqrt(q @ q)  # makes sure the quaternion is normalized
+
+    return q
+
+
+def ellipsoid_from_tensor(tensor, scale=1.0, force_positive=False):
+    """
+    Returns an ellipsoid (semiaxes + quaternion) representation of
+    a positive definite tensor (e.g. a polarizability), in the form
+    required by the chemiscope input.
+
+    :param tensor: a positive-definite tensor (3x3 or 6 array)
+    :param scale: conversion from the units of the tensor to the units of the atomic positions (usually Å)
+    :param force_positive: takes the absolute value of eigenvalues, to handle non-positive tensors
+    """
+
+    tensor = np.asarray(tensor) * (scale**2)
+    if len(tensor.flatten()) == 9:
+        matrix = tensor.reshape((3, 3))
+    elif len(tensor) == 6:
+        matrix = np.array(
+            [
+                [tensor[0], tensor[3], tensor[4]],
+                [tensor[3], tensor[1], tensor[5]],
+                [tensor[4], tensor[5], tensor[2]],
+            ]
+        )
+    # Compute the eigenvalues and eigenvectors
+    eigenvalues, eigenvectors = np.linalg.eigh(matrix)
+
+    if force_positive:
+        eigenvalues = np.abs(eigenvalues)
+
+    # The lengths of the principal axes are the square roots of the eigenvalues
+    old_settings = np.seterr(invalid="raise")
```

Comment:
Why are you bypassing this?

Reply:
I'm not bypassing it. By default it's a warning, and I'm turning it to an error.